node/subsystem/src/jaeger.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Jaeger integration.
+//!
+//! See <https://www.jaegertracing.io/> for an introduction.
+//!
+//! The easiest way to try Jaeger is:
+//!
+//! - Start a docker container with the all-in-one docker image (see below).
+//! - Open your browser and navigate to <https://localhost:16686> to acces the UI.
+//!
+//! The all-in-one image can be started with:
+//!
+//! ```not_rust
+//! podman login docker.io
+//! podman run -d --name jaeger \
+//!  -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
+//!  -p 5775:5775/udp \
+//!  -p 6831:6831/udp \
+//!  -p 6832:6832/udp \
+//!  -p 5778:5778 \
+//!  -p 16686:16686 \
+//!  -p 14268:14268 \
+//!  -p 14250:14250 \
+//!  -p 9411:9411 \
+//!  jaegertracing/all-in-one:1.21
+//! ```
+//!
+
+use polkadot_primitives::v1::{Hash, PoV, CandidateHash};
+use parking_lot::Mutex;
+use std::sync::Arc;
+use std::result;
+pub use crate::errors::JaegerError;
+
+
+lazy_static::lazy_static! {
+	static ref INSTANCE: Mutex<Jaeger> = Mutex::new(Jaeger::None);
```

Comment:
I would suggest using `parking_lot::RwLock` then. The only write is on initialization, everything else is a read.
We could use `read_recursive()` to avoid deadlock when used in an async context.

Reply:
Hmm, actually we also need a write for `add_string_tag` for peer_ids, but I'm not if our usage is correct, we seem to use the same key `peer-id`.